File: python-algorithm/ds_4_linkedlist/linkedlist_3_two_points/L160.cpp
//
// Created by LiHeng on 2020/4/19.
//
#include <iostream>
using namespace std;

//  Definition for singly-linked list.
struct ListNode {
    int val;
    ListNode *next;
    ListNode(int x) : val(x), next(NULL) {}
};

// 题目要求不能修改原来的链表结构，不能占用额外空间，因此反转链表和哈希查找法不符合要求
class Solution {
public:
    ListNode *getIntersectionNode(ListNode *headA, ListNode *headB) {
        ListNode* ha = headA;
        ListNode* hb = headB;

        while(ha != hb){ // 如果不相交，二者各走一遍对方的路，最终会同时还走向对方的终点都是空也会退出循环
            ha = (ha == NULL? headB:ha->next);
            hb = (hb == NULL? headA:hb->next);
        }
        return ha;
    }
};

File: python-algorithm/ds_4_linkedlist/linkedlist_3_two_points/L160.py
#--coding:utf-8--
'''
@Time   : 2020/8/14
@Author : Heng Li
@Email  : liheng_lulia@163.com
'''

# 另一种办法是各自遍历一次得到两者长度差值，然后快慢指针相遇
# Definition for singly-linked list.
class ListNode:
    def __init__(self, x):
        self.val = x
        self.next = None

class Solution:
    def getIntersectionNode(self, headA: ListNode, headB: ListNode) -> ListNode:
        node1, node2 = headA, headB
        while node1 != node2 :
            node1 = headB if node1 == None else node1.next
            node2 = headA if node2 == None else node2.next
        return node1